File: app/src/main/java/com/example/chris/atp_music_player/ui/SongListFragment.java
package com.example.chris.atp_music_player.ui;

import android.view.View;

import com.example.chris.atp_music_player.R;
import com.example.chris.atp_music_player.adapters.SongListAdapter;
import com.example.chris.atp_music_player.models.Song;
import com.example.chris.atp_music_player.provider.MusicProvider;

import java.util.List;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class SongListFragment extends ATPDataFragment {

    public static SongListFragment newInstance() {
        return new SongListFragment();
    }

    @Override
    protected void loadData() {
        MusicProvider musicProvider = new MusicProvider(getActivity());
        musicProvider.getAllSongs().subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread()).subscribe(new Subscriber<List<Song>>() {
            @Override
            public void onCompleted() {
            }

            @Override
            public void onError(Throwable e) {
            }

            @Override
            public void onNext(List<Song> songs) {
                if (songs.isEmpty()) {
                    mErrorView.setText(R.string.err_songs);
                    mRecyclerView.setVisibility(View.GONE);
                    mErrorView.setVisibility(View.VISIBLE);
                } else {
                    mRecyclerView.setAdapter(new SongListAdapter(getActivity(), songs));
                }
            }
        });
    }

}

File: app/src/main/java/com/example/chris/atp_music_player/ui/SongListFragment.kt
package com.example.chris.atp_music_player.ui

import android.view.View
import com.example.chris.atp_music_player.R
import com.example.chris.atp_music_player.SongRequestModel
import com.example.chris.atp_music_player.adapters.SongListAdapter
import com.example.chris.atp_music_player.playback.MediaProvider
import com.example.chris.atp_music_player.playback.SongRequest
import kotlinx.android.synthetic.main.no_data.*
import kotlinx.android.synthetic.main.song_list.*
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class SongListFragment : ATPListFragment() {
    override fun loadData() {
        MediaProvider(context)
                .getAllSongs()
                .toList()
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ songs ->
                    if (songs.isEmpty()) {
                        error.setText(R.string.err_songs)
                        recycler_view.visibility = View.GONE
                        error.visibility = View.VISIBLE
                    } else {
                        recycler_view.adapter = SongListAdapter(songs) { songs, integer ->
                            SongRequest.post(SongRequestModel(songs, integer, true))
                        }
                    }
                })
    }
}
